### ISSUE TITLE
Decimal and double literal type casting issues

### ISSUE THREAD

**Issue by yallie** [jayonsoftware/apexsharp#14]
C# decimal is incompatible with double literals and vice versa, double is incompatible with decimal literals:

```c#
decimal x = 123.45; // error
decimal x = 123.45m; // ok

double y = 123.45; // ok
double y = 123.45m; // error
```

The suffix cannot be put on all floating-point literals because that will render them incompatible with double variables, so we'll fix the issue with decimals just to get another issue with doubles. Perhaps this can be worked around via custom numeric types, like it's already done for Date/DateTime/Time.

Custom Decimal or Double class can be written so it accepts double or decimal literals with or without the "m" suffix:

```c#
decimal x = 123.45; // doesn't compile: native decimal isn't compatible with double
Decimal y = 123.45 // compiles: custom Apex.Decimal class is compatible

double z = 123.45m; // doesn't compile: native double isn't compatible with decimal
Double w = 123.45m; // compiles: custom Apex.Double class is compatible
```

Another option is to introduce custom numeric intermediate type. Say, `Number`, implicitly convertible to both decimal and double types. And cast all floating-point numeric literals to `Number` during conversion:

```c#
decimal x = (Number)123.45; // compiles because Number can be assigned to decimal
double y = (Number)123.45m; // compiles because Number can be assigned to double
```

But this way all C# expressions will be peppered with casts to `Number`. The casts will be removed when converted back to Apex:

```c#
decimal newPrice = oldPrice * (Number)1.5;
```

To sum up, I see two options of dealing with decimal/double casting issues:

* Custom Decimal and Double classes. Native decimal and double aren't used in C#.
* Intermediate Number type. Explicit type casts to Number on all C# floating-point literals.

So what way do we choose?
